### ISSUE TITLE
addFilter Crashes App when trying to filter requests

### ISSUE THREAD

**Issue by bradennapier** [elastic/apm-agent-nodejs#1385]
**Describe the bug**
A clear and concise description of what the bug is.  [Documentation indicates](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#apm-add-filter) that if we do not return a payload a given request will not be sent to the apm server.  This is true in some cases but not all.

```
The filter function is synchronous and should return the manipulated payload object. If a filter function doesn’t return any value or returns a falsy value, the remaining filter functions will not be called and the payload will not be sent to the APM Server.
```

There is also an issue with the `sampled` property not being the same everywhere which is confusing.

See https://github.com/elastic/apm/issues/151 for more context

**To Reproduce**

`apm-agent` version is `2.17.0`

In this case the app will crash any time that an error occurs within the request. Below is the general steps to make the error occur and details along the way

```
/**
 * Filter to only send sampled results to the APM Server
 */
apm.addFilter(payload => {
  console.log('\n--- Filter ---\n');
  console.log(JSON.stringify(payload, null, 2));

  if (typeof payload.sampled === 'boolean' && !payload.sampled) {
    /*
      CASE 1

      When this occurs the data result looks similar to below.  
      Returning here when not sampling works as expected and does
      not produce any errors.
    */
    return;
  }

  if (
    typeof payload.transaction?.sampled === 'boolean' &&
    !payload.transaction.sampled
  ) {
    /*
      CASE 2

      When an error occurs, the `transaction.sampled` is where
      the data is, but regardless, if this filter ever returns
      undefined it will cause the app to crash completely.
    */
    return;
  }

  return payload;
});
```

## [CASE 1]: First Filter Works

It first provides the given payload.  If I return undefined for this then it does not crash the application.  However, this is a tiny data payload, my goal is to completely ignore anything that is not sampled. As otherwise ES will cost me many tens of thousands of dollars per month to run with how it is currently built.

This appears to just be used to add metrics and do some math - those I am actually fine with... its the indexing of EVERY single request that is the problem.

```
{
  "id": "566594a22f516683",
  "trace_id": "fc9d048d55f21ffbebb1a2d44648b7d3",
  "name": "GET /example",
  "type": "request",
  "duration": 94.044,
  "timestamp": 1569358244043041,
  "result": "HTTP 2xx",
  "sampled": false,
  "sync": false,
  "span_count": {
    "started": 0
  }
}
```

## [CASE 2] `transaction.type` of `request` Crashes

In this case, returning `undefined` crashes the app with the following stacktrace.  

> **Note:** I modified the payload below to remove extraneous payload - but its a normal http request/exception case.  

```javascript
{
  "exception": {
    "message": "Failed to validate request [1]",
    "type": "UnauthorizedError",
    "code": "Unauthorized",
    "attributes": {
      "jse_shortmsg": "Failed to validate request [1]",
      "message": "Failed to validate request [1]"
    },
    "handled": true
  },
  "id": "5b6003f09be05fc325400882513f357c",
  "parent_id": "d4106918fb0b26a6",
  "trace_id": "dbffee8786a51aace74e9efde9ab91ba",
  "timestamp": 1569358223470000,
    "custom": {},
  "transaction_id": "d4106918fb0b26a6",
  "transaction": {
    "type": "request",
    "sampled": false
  }
}
```

```
  TypeError: Cannot read property 'id' of undefined
  
  - agent.js:391 send
    [elastic-apm-node]/lib/agent.js:391:68
  
  - agent.js:383 prepareError
    [elastic-apm-node]/lib/agent.js:383:7
  
  - agent.js:299 
    [elastic-apm-node]/lib/agent.js:299:7
  
  - parsers.js:81 
    [elastic-apm-node]/lib/parsers.js:81:7
  
  - index.js:20 
    [after-all-results]/index.js:20:25
  
  - next_tick.js:61 process._tickCallback
    internal/process/next_tick.js:61:11
```

**Comment by bradennapier** [comment]
Quick look at code reveals the problem pretty quickly.

```javascript
 error = agent._errorFilters.process(error)

    if (!error) {
      agent.logger.debug('error ignored by filter %o', { id: error.id })
      if (cb) cb(null, error.id)
      return
    }
```

If `!error` then you probably shouldn't check for `error.id` ;-) -- guess maybe returning `null` should fix this in the meantime?

**Comment by Qard** [comment]
Good catch, I'll get that fixed.

In the mean time, you could use `agent.addTransactionFilter(...)` and `agent.addSpanFilter(...)` to only filter transactions and spans. Errors _should_ be infrequent anyway, unless there's something wrong with your system, in which case you probably want to know about it. 😅

**Comment by bradennapier** [comment]
@Qard errors make up a huge % of our total requests (many millions per hour).  This is our general public API and errors are quite normal.  

They may be rate limit errors or any other internal error and are critical for tracking, but since apm can't handle our scale without us spending almost 2x what we spend on the API itself we need to sample them.

**Comment by Qard** [comment]
We're in the middle of preparing a major release, so I'll see what I can do to get something out in 2.x quickly. 😅

**Comment by axw** [comment]
@bradennapier thanks very much for the super detailed bug description and analysis, this is really helpful!

**Comment by watson** [comment]
I have a fix ready for this in #1389 - The reason why our tests didn't catch this previously was that our tests returned `false` instead of `undefined` from within the filter. And `(false).id` is valid code whereas `(undefined).id` throws.

Until this lands, a quick fix would, therefore, be to replace your `return` statements with `return false`. Obviously this would mean that the `id` logged and provided in the callback would be `undefined` itself, but at least the code wouldn't throw.

I've updated the tests as well to test all falsy values.

We'll let you know as soon as the fix lands.

**Comment by bradennapier** [comment]
![image](https://user-images.githubusercontent.com/15365418/65625919-b79d2e80-df9a-11e9-8b69-9d04c7573a46.png)

It is amazing what sampling can do for a server ;-).

Haven't checked everything fully yet but that does appear to fix it for now.

**Comment by watson** [comment]
A fix for this has now been released in v2.17.1. Thanks a lot again for reporting this issue 💯
